### ISSUE TITLE
datapath: Emit trace events in XDP for debug purpose

### ISSUE THREAD

**Issue by brb** [cilium/cilium#19703]
Currently, to debug XDP we rely on the PCAP recorder. Unfortunately, the recorder doesn't work with Hubble. I.e., the recorder flows cannot be shown in Hubble UI.

For debugging purpose, we could introduce a new debug mode (e.g., `--debug-verbose=xdp`) which would enable emitting of cilium flow events to the perf ring buffer.

Bonus points, if we could enable this during runtime via `cilium config`.

**Comment by dylandreimerink** [comment]
@brb could you elaborate on this a little more?

I have taken a look into the current flow of event data. We have 3 kinds of messages which sort of do the same thing: 
* Debug capture events 
* Trace events
* Capture record events

The debug capture and trace events are [parsed](https://github.com/cilium/cilium/blob/4f1316538ba6a638221021b3da4c5caeb2b35e72/pkg/hubble/parser/threefour/parser.go#L135) into flows but the capture record events are not.

So my questions are:
1. Wouldn't we rather create flow data from capture record events? Since we have that data when recording something.
2. If we still want to use add a generic "capture all" mode, should we use debug capture events or trace events? (not sure when to use one over the other)
3. Do we just want to record all traffic, including traffic that might be dropped by the prefilter? Or just the traffic we actually send to the nodeport entrypoint?

**Comment by brb** [comment]
@dylandreimerink Good questions. 

1. The issue is about enabling `cilium_dbg` and `send_trace_notify` in `bpf/bpf_xdp.c`. Converting captured record events (aka recorded PCAPs) into parsable ones by Hubble is out of scope of this issue, as it might involve a fair amount of work.
2. Once we have both enabled, then let's add two `send_trace_notify()` - one before the LB selection (i.e., before `nodeport_lb{4,6}()`) and one after (i.e., after all packet mangling has happened).
3. For now I would be interested only in NodePort traffic. No need to consider the prefilter, as it might be used only by a few.

**Comment by dylandreimerink** [comment]
@brb I have [this](https://github.com/dylandreimerink/cilium/commit/2ecf4f8604c524ed2c6d3ba77fdebddb6922ce46) so far, not quite happy yet, I wanted to discuss the folowing:

1. Do we want a separate `--debug-verbose=xdp`? Doesn't seem necessary since other trace functions just aggregate if not monitored
2. Technically the term `FROM_NETWORK` applies more to the XDP programs than the TC programs. Should I remove the `FROM_NETWORK` traces and use this trace_point in the XDP "from-network" program? Or should we introduce a new trace_point enum? (i.e. `FROM_XDP`?)
3. If we define 2 new enums(pre and post mangle), do we want metrics for those as well? or just discard metrics like with `FROM_LXC` and `TRACE_{FROM,TO}_PROXY` ?
4. Should we aggregate these bytes on `METRIC_INGRESS`+`REASON_PLAINTEXT` or will that mess up the count since after the XDP program the same packet will also be processed by the TC program?

**Comment by brb** [comment]
When I created the issue I was thinking about enabling Cilium traces on XDP with the least amount of work so that they can be shown by Hubble. The initial thought was to add `send_trace_notify()` protected with `#ifdef ENABLE_XDP_TRACING` or so.

1. I think the approach with the aggregation would only work if we introduce a new trace point. I would suggest to not introduce it considering the motivation above.
2. `FROM_NETWORK` is OK.
3. Let's not introduce new enums (see 1.)
4. I wouldn't be concerned too much about aggregation, as this feature would be used only for debugging purpose.

**Comment by dylandreimerink** [comment]
A side effect of using `FROM_NETWORK` is also that we see the same flow 2 or 3 times depending on if a packet gets mangled or not. It looks strange in the monitor output, though I suppose that it does make sure the flows get to Hubble.

```
xx drop (Stale or unroutable IP) flow 0xb9a67965 to endpoint 0, , identity host->unknown: [fd00::c]:52390 -> [fd04::1:5179]:4240 tcp ACK, FIN
-> network flow 0xc8b0e839 , identity unknown->unknown state unknown ifindex 0 orig-ip 0.0.0.0: [fd01::c]:39504 -> [fd01::b]:4240 tcp ACK
<- network flow 0x0 , identity unknown->unknown state unknown ifindex enp0s8 orig-ip 0.0.0.0: [fd01::b]:4240 -> [fd01::c]:39504 tcp ACK
<- network flow 0x0 , identity unknown->unknown state unknown ifindex enp0s8 orig-ip 0.0.0.0: [fd01::b]:4240 -> [fd01::c]:39504 tcp ACK
<- network flow 0x0 , identity unknown->unknown state unknown ifindex enp0s8 orig-ip 0.0.0.0: [fd01::b]:4240 -> [fd01::c]:39504 tcp ACK
```

It is something we can address now, in a future version or not at all and just add a note somewhere

**Comment by brb** [comment]
As it's going to be used only for debugging purpose, I wouldn't be too concerned about the duplicate events.

**Comment by brb** [comment]
@knfoo The PR https://github.com/cilium/cilium/pull/19797 makes it possible via `kubectl exec -ti -n kube-system $CILIUM_POD_NAME -- cilium config XDPTracing=enabled`.

**Comment by knfoo** [comment]
@brb Yeah makes sense and  that is a good thing to be able to do that.
With this implementation is it limited to users with rights to access cilium pods which is very limited in our case.
We will be able to create tooling around this to allow "regular" users to enable this. But a more clean implementation would be to be able to expose it via object annotations like you have with the other L7 viability in cilium

**Comment by brb** [comment]
> But a more clean implementation would be to be able to expose it via object annotations like you have with the other L7 viability in cilium

I am a bit reluctant of adding a way to configure cilium-agent's via CiliumNode object annotations in #19797. Mainly, because we haven't done it yet, and I think that we need a broader discussion on how to do it properly.

**Comment by knfoo** [comment]
> I am a bit reluctant of adding a way to configure cilium-agent's via CiliumNode object annotations in https://github.com/cilium/cilium/pull/19797. Mainly, because we haven't done it yet, and I think that we need a broader discussion on how to do it properly.
Fair enough. Should I create a new issue for that or should that be tracked in this issue then ?

**Comment by brb** [comment]
Yep, creating an issue would work (just please refer to the PR so that folks would have enough context).

**Comment by brb** [comment]
MBOI https://github.com/cilium/pwru/pull/339.
